design/baremetal-operator/disable-introspection-intereface.md

```diff
@@ -0,0 +1,96 @@
+# disable-inspection API
+
+## Status
+
+implementable
+
+## Summary
+
+A declarative API is proposed to disable inspection of a BareMetalHost.
+
+## Motivation
+
+Related to the work to enable booting of a
+[LiveImage](https://github.com/metal3-io/metal3-docs/pull/150),
+there is the requirement to optionally disable inspection on initial
+registration of a BareMetalHost (so that the live image can boot
+more quickly, ref user stories below).
+
+### Goals
+
+- A declarative API to disable inspection on BMH registration
+- The option to update status data with a status annotation when inspection is
+  performed by some means other than the regular BMH flow
+- An interface that's consistent with existing proposals, in particular the
+  [re-inspection interface](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/re-introspection-interface.md)
+
+### Non-Goals
+
+- There are no plans for any mechanism to trigger or consume data from any out
+  of band inspection process, other than providing data via the existing
+  status annotation.
+
+## Proposal
+
+To align with the re-inspection API proposal, the `inspect.metal3.io`
+annotation will be reused, but with an optional additional suffix, similar
+to that described for the [reboot API](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/reboot-interface.md)
+
+The optional `inspect.metal3.io/disable` annotation will be used to describe
+the situation where we wish to disable the default inspection behavior.
+
+When the BMO finds this annotation, it will move directly from `Registering`
+[state](https://github.com/metal3-io/baremetal-operator/blob/master/docs/BaremetalHost_ProvisioningState.png)
+to `Ready` and not perform any inspection.
+
+In this case (similar to the reboot interface), when the suffixed form of
+`inspect.metal3.io/disable` is found, it will never be removed by the BMO.
+
+If both inspect.metal3.io and inspect.metal3.io/disable forms are in specified,
+the `inspect.metal3.io/disable` form will take priority and the
+`inspect.metal3.io` annotation will be ignored.
+
+When `baremetalhost.metal3.io/status` is updated, any data should be
+consumed even in the `Ready` or `Provisioned` state, so that in the
+case where inspection data is collected via a live-image we can still
+store that data in the BMH status field.
```

Comment:
This will change how we manage deleting the annotation. Today we delete it if it is present and the status fields are not nil. If we're going to allow updates, we will need to be more careful.

Which status fields do we want to allow someone to update this way? The current annotation encodes the entire status block, including provisioning state. As a one-time thing, that's OK. I'm not sure we want to give anyone with permission to add an annotation to a host permission to overwrite its "internal" provisioning status.

If we only care about allowing updates to the hardware details, perhaps we want a new annotation for that? Or we want at least to specify that only the hardware details will be updated?

We should also specify whether the settings from the annotation are merged with or overwrite the existing settings. I think we're going to want them to overwrite?

Reply:
Oh I didn't realize the entire status could be overwritten, so yes we'll need to be more careful and I think we only care about updating the hardware details.

Perhaps it should be a new suffix e.g a path like `baremetalhost.metal3.io/status/hardware` - and that is allowed to be updated (either unconditionally, or when inspection is disabled?)

Reply:
I feel like disabling introspection and updating the hardware details are slightly orthogonal. So, I would use a new annotation (to be bikeshedded later) for the hardware details and apply that unconditionally. Then I would add the annotation to disable inspection to avoid that part of the workflow.

The implementation of the hardware details update probably belongs in the controller near the place where we update the status based on the annotation now.

For the implementation of the disable annotation, I think we could still go through the same states in the state machine, and just have the handler for the inspection state skip its work if the annotation is present. That seems simpler than adding a new transition? I could go either way on that part.

Reply:
Ok so perhaps I should remove the status annotation content from this PR and create another separate proposal?

They're both related to the same use-cases, but I agree the implementation can be approached independently.

I also don't have a strong opinion on a new state vs skipping inspection in the existing state - happy to take your guidance on which is likely to be simpler as I've not looked that closely at the code in the state machine yet :)

Reply:
> Ok so perhaps I should remove the status annotation content from this PR and create another separate proposal?

I think we can reframe this proposal as "support external hardware inspection" and talk about both here.

> 
> They're both related to the same use-cases, but I agree the implementation can be approached independently.
> 
> I also don't have a strong opinion on a new state vs skipping inspection in the existing state - happy to take your guidance on which is likely to be simpler as I've not looked that closely at the code in the state machine yet :)

@zaneb or @andfasano would be better judges these days, to be honest. :-)

Reply:
From the given description, I'd give my preference to the inspection skipping during the `Inspecting` state, rather than adding a new state - there's not enough "critical mass" to justify a new state IMHO, and the skip implementation would be pretty straightforward.

Reply:
Ok I retitled the proposal, and split the disable/update parts in the proposal section.  Also mentioned that we'll skip inspecting in the `Inspecting` state.

Reply:
I think the discussion got garbled a little bit - Doug talked about adding a new state _transition_ (i.e. from Registering->MatchProfile, skipping Inspecting), not adding a new state (which I agree we don't want).
Either doing the new transition (check annotation in `NeedsInspection()`) or just handling it during the `Inspecting` state (check annotation in `actionInspecting()`) is easy enough, so I don't really care either.